README.md
# parse-numeric-date

Micro library to parse numeric based date inputs
~4kb minified/~1kb gzip

## Mission Statement

Date entry is hard for users. Date pickers, including the native one, aren't always the best for accessibility. 

In his article ["Maybe You Don't Need a Date Picker"](https://adrianroselli.com/2019/07/maybe-you-dont-need-a-date-picker.html#Messaging), Adrian Roselli proposes using a text field which users enter a date into, some client side does some parsing and then some feedback is presented to the user.

It is fairly straight forward to parse a known date format and convert to ISO format. But, what if users don't enter the date in your known format?

Well, that is what `parse-numeric-date` attempts to solve.

## Demo

[https://area17.github.io/parse-numeric-date/](https://area17.github.io/parse-numeric-date/)

## Usage

Firstly, install via npm:

```
$ npm install @area17/parse-numeric-date
```

Import into your JavaScript by:

```
import parseNumericDate from '@area17/parse-numeric-date'
```

And then in your JavaScript:

```
let isoString = parseNumericDate('29/12/25'); // 2025-12-29
```

If you want to have manual control over the locale:

```
let isoString = parseNumericDate('29/12/25', {
    locale: 'en-GB' // will favor d/m/y over m/d/y
}); // 2025-12-29
```

## Examples of what `parse-numeric-date` can parse

* 2022-02-10
* 2022.02.10
* 2022/02/10
* 2022-2-10
* 2022-2-2
* 10 02 2022
* 1 12 2022
* 122922
* 090322
* 9322
* 29122022
* 010122
* 12/29/2022
* 12/29/22
* 01-01-2022
* 01-01-22
* 1-1-22
* 1-1-2022
* ۱۴۰۰/۱۱/۲۱
* 2022年12月13日
* `13. 12. 2022`
* 13‏/12‏/2022.

## Notes

This `parse-numeric-date` works with a big assumption:

*Dates closer to today are better than dates years into the future*

Date pickers for hotels, restaurants or events tend to be for things happening this year or next year. Dates of birth tend to happen from now to 100 years in the past. Retirement ages, mortgage completions tend to happen from now to 100 years in the future. 

If the year difference is larger than 100, its probably useless.
eg: a 29 in the string appear as 2922 rather than 29th of a month, 2022

## Limitations

`parse-numeric-date` can't take natural input such as "December 12, 2022" - it is purely for parsing numeric inputs. Another assumption here is that such date entry, although natural in speech, isn't so natural to type - especially if the keyboard has been switched to a numeric keyboard on the users device.

If you want such parsing, perhaps you need:
* [`anyDateParser`](https://github.com/kensnyder/any-date-parser)
* [`parse-dob`](https://github.com/HenrikJoreteg/parse-dob)
* [`chrono-node`](https://github.com/wanasit/chrono)

Both are excellent, but can't parse numeric dates quite as fully as `parse-numeric-date`.

## Comparison with other libraries

You can see a comparison between `parse-numeric-date`, [`anyDateParser`](https://github.com/kensnyder/any-date-parser), [`parse-dob`](https://github.com/HenrikJoreteg/parse-dob) and [`chrono-node`](https://github.com/wanasit/chrono) over on Codepen: [Date parse testing](https://codepen.io/13twelve/pen/rNrRoLB)

`parse-dob` by [@HenrikJoreteg](https://github.com/HenrikJoreteg) was a great inspiration, we think that forking `parse-dob` and removing the previous date restriction would be very interesting.

### Tests

Written using [Jest](https://jestjs.io/) with tests in `/test/`. 
To run:

```Shell
$ npm run test
```

## Credits

Thanks to [@Mohsen Alyafei](https://github.com/MohsenAlyafei) for his Persian to Gregorian conversion function on [Stackoverflow](https://stackoverflow.com/questions/71421825/how-to-convert-persian-jalali-dates-to-other-18-calendar-dates-in-javascript-w).

## Code of Conduct

AREA 17 is dedicated to building a welcoming, diverse, safe community. We expect everyone participating in the AREA 17 community to abide by our [Code of Conduct](CODE_OF_CONDUCT.md). Please read it. Please follow it.


changelog.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## 0.0.9
* Two year cut off is now dynamic and not fixed
* Returns null and a `console.log` on none valid entry 

## 0.0.8
Refactored to make it easier to add ambiguous combos - removes a few bytes.

## 0.0.7
Add parsing date types:
* ۱۴۰۰/۱۱/۲۱
* 2022年12月13日
* `13. 12. 2022`
* 13‏/12‏/2022.

## 0.0.6
* Fix trapping no parsed dates when determining highest score
* Add [Jest](https://jestjs.io/) tests: `npm run test`

## 0.0.5
Readme updates for [Date parse testing](https://codepen.io/13twelve/pen/rNrRoLB) comparison with `chrono-node` update.

## 0.0.4
Add parsing of 5 and 7 digit non deliminated dates

## 0.0.3
Add debug dev task

## 0.0.2
Make options, optional 🤦🏻‍♂️

## 0.0.1
Initial release


test/parseNumericDate.test.js
import parseNumericDate from '../src/index';

const tests = [
  ['2022-02-10', '2022-02-10'],
  ['2022.02.10', '2022-02-10'],
  ['2022/02/10', '2022-02-10'],
  ['10 02 2022', '2022-02-10'],
  ['1 12 2022', '2022-12-01'],
  ['2022-2-10', '2022-02-10'],
  ['2022-2-2', '2022-02-02'],
  ['122922', '2022-12-29'],
  ['291222', '2022-12-29'],
  ['29122022', '2022-12-29'],
  ['12292022', '2022-12-29'],
  ['010122', '2022-01-01'],
  ['220210', '2022-02-10'],
  ['12/29/22', '2022-12-29'],
  ['2010-02-01', '2010-02-01'],
  ['01-01-2022', '2022-01-01'],
  ['01-01-22', '2022-01-01'],
  ['1-1-22', '2022-01-01'],
  ['1-10-2022', '2022-10-01'],
  ['1-10-22', '2022-10-01'],
  ['122982', '1982-12-29'],
  ['291282', '1982-12-29'],
  ['29121982', '1982-12-29'],
  ['12291982', '1982-12-29'],
  ['010182', '1982-01-01'],
  ['9382', '1982-03-09'],
  ['931982', '1982-03-09'],
  ['090382', '1982-03-09'],
  ['09031982', '1982-03-09'],
  ['2023-02-10', '2023-02-10'],
  ['2023.02.10', '2023-02-10'],
  ['2023/02/10', '2023-02-10'],
  ['10 02 2023', '2023-02-10'],
  ['1 12 2023', '2023-12-01'],
  ['2023-2-10', '2023-02-10'],
  ['2023-2-2', '2023-02-02'],
  ['122923', '2023-12-29'],
  ['291223', '2023-12-29'],
  ['29122023', '2023-12-29'],
  ['12292023', '2023-12-29'],
  ['010123', '2023-01-01'],
  ['230210', '2023-02-10'],
  ['12/29/23', '2023-12-29'],
  ['2023-02-01', '2023-02-01'],
  ['01-01-2023', '2023-01-01'],
  ['01-01-23', '2023-01-01'],
  ['1-1-23', '2023-01-01'],
  ['1-10-2023', '2023-10-01'],
  ['1-10-23', '2023-10-01'],
  ['2024-02-10', '2024-02-10'],
  ['2024.02.10', '2024-02-10'],
  ['2024/02/10', '2024-02-10'],
  ['10 02 2024', '2024-02-10'],
  ['1 12 2024', '2024-12-01'],
  ['2024-2-10', '2024-02-10'],
  ['2024-2-2', '2024-02-02'],
  ['122924', '2024-12-29'],
  ['291224', '2024-12-29'],
  ['29122024', '2024-12-29'],
  ['12292024', '2024-12-29'],
  ['010124', '2024-01-01'],
  ['240210', '2024-02-10'],
  ['12/29/24', '2024-12-29'],
  ['2024-02-01', '2024-02-01'],
  ['01-01-2024', '2024-01-01'],
  ['01-01-24', '2024-01-01'],
  ['1-1-24', '2024-01-01'],
  ['1-10-2024', '2024-10-01'],
  ['1-10-24', '2024-10-01'],
  ['1884-02-10', '1884-02-10'],
  ['1884.02.10', '1884-02-10'],
  ['1884/02/10', '1884-02-10'],
  ['10 02 1884', '1884-02-10'],
  ['1 12 1884', '1884-12-01'],
  ['12 29 1884', '1884-12-29'],
  ['1884-2-10', '1884-02-10'],
  ['1884-2-2', '1884-02-02'],
  ['1884-02-01', '1884-02-01'],
  ['01-01-1884', '1884-01-01'],
  ['1-10-1884', '1884-10-01'],
  ['۱۴۰۰/۱۱/۲۱', '2022-02-10'],
  ['13. 12. 2022', '2022-12-13'],
  ['13‏/12‏/2022', '2022-12-13'],
  ['2022年12月13日', '2022-12-13'],
];

const illogical = [
  ['0000-00-00', null],
  ['00000000', null],
  ['11', null],
  ['today', null],
  ['⛄️', null],
];

describe('parseNumericDate', () => {
    it('exists', () => {
        expect(typeof parseNumericDate).toBe('function');
    });

    it('it correctly transforms dates', () => {
        tests.forEach((test, index) => {
          const d = parseNumericDate(test[0]);
          const e = new Date(`${test[1]}T00:00:00.000+00:00`);
          expect(d).toEqual(e);

          const isDate = d instanceof Date;
          expect(isDate).toEqual(true);
        });
    });

    it('it returns null for illogical input', () => {
        illogical.forEach((test, index) => {
          const d = parseNumericDate(test[0]);
          const e = test[1];
          expect(d).toEqual(e);

          const isDate = d instanceof Date;
          expect(isDate).toEqual(false);
        });
    });
});


src/index.js
let locale = new Intl.NumberFormat().resolvedOptions().locale;
const today = new Date().setHours(0, 0, 0, 0);
const thisYear = new Date(today).getFullYear();
const yearInMs = 365.25 * 24 * 60 * 60 * 1000;
const twoYearsInMs = yearInMs * 2;
const hundredYearsInMs = yearInMs * 100;
const getLocaleOrder = () => {
  const dateStr = new Date(2000, 11, 13).toLocaleDateString(locale);
  const parts = [
    { name: 'd', index: dateStr.indexOf('13') },
    { name: 'm', index: dateStr.indexOf('12') },
    { name: 'y', index: dateStr.indexOf('2000') },
  ].sort((a, b) => a.index - b.index);
  return parts.map(item => item.name).join('-');
};

let localeOrder = getLocaleOrder();

// Month in JavaScript is 0-indexed (January is 0, February is 1, etc),
// but by using 0 as the day it will give us the last day of the prior
// month. So passing in 1 as the month number will return the last day
// of January, not February
const daysInMonth = (m, y) => {
  return new Date(y, m, 0).getDate();
};

const make4DigitYear = (y) => {
  if (y.length === 2) {
    let twoDigitCutoffYear = 50;
    let century = Math.floor(thisYear / 1000) * 1000;
    let year = century + parseInt(y, 10);
    let diff = Math.abs(year - (thisYear + twoDigitCutoffYear));
    return (year < thisYear + twoDigitCutoffYear) || diff > 100 ? year : year - 100;
  }
  return y;
}

/*
    persianToGregorian,
    adapted from: https://stackoverflow.com/questions/71421825/how-to-convert-persian-jalali-dates-to-other-18-calendar-dates-in-javascript-w

    credit to Mohsen Alyafei
    https://github.com/MohsenAlyafei
    https://twitter.com/maalyafei
*/
function persianToGregorian(year, month, day) {
  year = parseInt(year, 10);
  month = parseInt(month, 10);
  day = parseInt(day, 10);

  const dFormat = new Intl.DateTimeFormat('en-u-ca-persian',{
    dateStyle:'short',
    timeZone:'UTC'
  });

  let gregorianDate = new Date(Date.UTC(2000,month,day));

  gregorianDate = new Date(gregorianDate.setUTCDate(gregorianDate.getUTCDate() + 226867));

  const gregorianYear = gregorianDate.getUTCFullYear() - 2000 + year;
  gregorianDate = new Date(
    ((gregorianYear < 0) ? '-' : '+') +
    ('00000' + Math.abs(gregorianYear)).slice(-6) +
    '-' +
    ('0' + (gregorianDate.getUTCMonth() + 1)).slice(-2) +
    '-' +
    ('0' + (gregorianDate.getUTCDate())).slice(-2)
  );

  let [pM, pD, pY] = [...dFormat.format(gregorianDate).split('/')]
  let i = 0;

  gregorianDate = new Date(
    gregorianDate.setUTCDate(
      gregorianDate.getUTCDate() +
      Math.floor(
        (year * 365.25) +
        (month * 30.44) +
        day -
        (
          (pY.split(' ')[0] * 365.25) +
          (pM * 30.44) +
          (pD*1)
        )
      ) - 2
    )
  );

  while (i < 4) {
    [pM, pD, pY] = [...dFormat.format(gregorianDate).split('/')];
    if (parseInt(pD, 10) === day && parseInt(pM , 10) === month && parseInt(pY.split(' ')[0], 10) === year) {
      return gregorianDate;
    }
    gregorianDate = new Date(gregorianDate.setUTCDate(gregorianDate.getUTCDate() + 1));
    i++;
  }
  console.log('Invalid Persian date');
  throw new Error('Invalid Persian date');
}

const getDatePart = (dateStr, format, part) => {
    return dateStr.substring(format.indexOf(part), format.indexOf(part) + (format.match(new RegExp(part, 'g')) || []).length);
};

const splitDateString = (dateStr, format) => {
    const parts = [
      { name: 'd', index: format.indexOf('D') },
      { name: 'm', index: format.indexOf('M') },
      { name: 'y', index: format.indexOf('Y') },
    ].sort((a, b) => a.index - b.index);

    return {
        y: getDatePart(dateStr, format, 'Y'),
        m: getDatePart(dateStr, format, 'M'),
        d: getDatePart(dateStr, format, 'D'),
        order: parts.map(item => item.name).join('-'),
    }
};

const splitDates = (dateStr, formats) => {
    return formats.map(format => splitDateString(dateStr, format));
};

const generateDate = (y, m, d) => {
  // if year is 2 digits, guess 19XX or 20XX
  y = make4DigitYear(y);

  // convert to numbers
  y = parseInt(y, 10);
  m = parseInt(m, 10);
  d = parseInt(d, 10);

  // stop months being over 12
  m = m > 12 ? null : m;

  // stop day being over how many days are in the month
  d = d > daysInMonth(m, y) ? null : d;

  // pad month and day
  m = m < 10 ? `0${m}` : m;
  d = d < 10 ? `0${d}` : d;

  // test date
  let t = new Date(`${y}-${m}-${d}T00:00:00.000+00:00`);

  if (t instanceof Date && !isNaN(t)) {
    return t;
  } else {
    return null;
  }
};

const getMostProbableDate = (dates, dateStr) => {
  // make date strs, set order score
  // filter out nulls
  // order by score
  dates = dates
    .map(obj => {
      // make date
      obj.d = generateDate(obj.y, obj.m, obj.d);
      // make year a number
      obj.y = parseInt(make4DigitYear(obj.y), 10);
      // find how far away from this year the date is
      //obj.diff = Math.abs(new Date(today).getFullYear() - obj.y);
      obj.diff = Math.abs(today - obj.d);
      // if the date is in locale order, give it a higher score
      obj.score = localeOrder === obj.order ? 2 : 1;
      //
      // Assumption: dates closer to today are better
      // Why: date pickers for hotels, restaurants, events tend to
      // happen this year or next year. Dates of birth tend to happen
      // from now to 100 years in the past. Retirement ages, mortgage
      // completions tend to happen from now to 100 years in the future
      //
      // if the year difference is larger than 100, its probably useless
      // eg: a 29 made it 2922 rather than 29th of a month, 2022
      if (obj.diff > hundredYearsInMs) {
        obj.score--;
      }
      // year is really quite close to this year, probably more likely
      if (obj.diff < twoYearsInMs) {
        obj.score++;
      }

      return obj;
    })
    // remove any null dates from invalid dates being made
    .filter(obj => obj.d !== null)
    // sort by score, high to low
    .sort((a, b) => {
      return b.score - a.score;
    });

  // whats our highest score?
  const highestScore = dates.length ? dates[0].score : 0;

  // remove anything that hasn't got the lowest score
  // get just the dates
  // order by closest to today
  dates = dates
    // remove anything that doesn't match our high score
    .filter(obj => obj.score === highestScore)
    // sort by year distance, low to high
    .sort((a, b) => {
      return a.diff - b.diff;
    })
    .map(obj => obj.d);

  // return first date (highest score, lowest year difference)
  if (dates.length) {
    return dates[0];
  }
  return null;
};

const parseNumericDate = (dateStr, options) => {
  if (options?.locale) {
    locale = options.locale;
    localeOrder = getLocaleOrder();
  }

  let originalDateStr = dateStr;
  dateStr = dateStr.replace(/[۰-۹]/g, d => '۰۱۲۳۴۵۶۷۸۹'.indexOf(d))
    .replace(/[٠-٩]/g, d => '٠١٢٣٤٥٦٧٨٩'.indexOf(d)).trim();
  let dates = [];

  // input probably = ISO YYYY-MM-DD or YYYY-M-D
  let matches = [
    ...dateStr.matchAll(/([0-9]{4})[^0-9]{1,}([0-9]{1,2})[^0-9]{1,}([0-9]{1,2})/g)
  ];
  if (matches.length > 0) {
    if (originalDateStr.match(/[۰-۹]/g)) {
        return persianToGregorian(matches[0][1], matches[0][2], matches[0][3]);
    } else {
        return generateDate(matches[0][1], matches[0][2], matches[0][3]);
    }
  }

  // input doesn't follow ISO YYYY-MM-DD
  // lets check for DD-MM-YYYY (or MM-DD-YYYY)
  matches = [
    ...dateStr.matchAll(/([0-9]{1,2})[^0-9]{1,}([0-9]{2})[^0-9]{1,}([0-9]{4})/g)
  ];
  if (matches.length > 0) {
    dates.push({
      y: matches[0][3],
      m: matches[0][2],
      d: matches[0][1],
      order: 'd-m-y'
    });

    dates.push({
      y: matches[0][3],
      m: matches[0][1],
      d: matches[0][2],
      order: 'm-d-y'
    });
    //
    return getMostProbableDate(dates, dateStr);
  }

  // lets check for DD-MM-YY (or MM-DD-YY), YY-MM-DD
  // and D-M-YY, M-DD-YY
  matches = [
    ...dateStr.matchAll(
      /([0-9]{1,2})[^0-9]{1,}([0-9]{1,2})[^0-9]{1,}([0-9]{1,2})/g
    )
  ];
  if (matches.length > 0) {
    dates.push({
      y: matches[0][3],
      m: matches[0][2],
      d: matches[0][1],
      order: 'd-m-y'
    });

    dates.push({
      y: matches[0][3],
      m: matches[0][1],
      d: matches[0][2],
      order: 'm-d-y'
    });

    dates.push({
      y: matches[0][1],
      m: matches[0][2],
      d: matches[0][3],
      order: 'y-m-d'
    });
    //
    return getMostProbableDate(dates, originalDateStr);
  }

  // input has no spaces between dates
  // very much guessing now

  matches = [
    ...dateStr.matchAll(
      /[0-9]{1,}/g
    )
  ];
  if (matches.length > 0) {
      // could be YYYYMMDD, DDMMYYYY, MMDDYYYY
      if (dateStr.length === 8) {
        dates = splitDates(dateStr, ['YYYYMMDD', 'DDMMYYYY', 'MMDDYYYY']);
      }

      // could be YYYYMMD, YYYYMDD,
      // or DDMYYYY, MDDYYYY, DMMYYYY, MMDYYYY
      if (dateStr.length === 7) {
        dates = splitDates(dateStr, ['YYYYMMD', 'YYYYMDD', 'DDMYYYY', 'MDDYYYY', 'DMMYYYY', 'MMDYYYY']);
      }

      // or could be YYMMDD, DDMMYY, MMDDYY
      // or worse, could be YYYYMD, DMYYYY, MDYYYY
      if (dateStr.length === 6) {
        dates = splitDates(dateStr, ['YYMMDD', 'DDMMYY', 'MMDDYY', 'YYYYMD', 'DMYYYY', 'MDYYYY']);
      }

      // could be YYMMD, YYMDD,
      // or DDMYY, MDDYY, DMMYY, MMDYY
      if (dateStr.length === 5) {
        dates = splitDates(dateStr, ['YYMMD', 'YYMDD', 'DDMYY', 'MDDYY', 'DMMYY', 'MMDYY']);
      }

      // could be YYMD, DMYY, MDYY
      if (dateStr.length === 4) {
        dates = splitDates(dateStr, ['YYMD', 'DMYY', 'MDYY']);
      }

      //
      return getMostProbableDate(dates, dateStr);
   }

   console.log('parseNumericDate - unknown format:', dateStr);
   return null;
};

export default parseNumericDate;
